Bias Correction Implementation

edmundw314 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#41845 Bias Correction Implementation**

Differential Revision: [D22661503](https://our.internmc.facebook.com/intern/diff/D22661503)

edmundw314 committed:
Bias Correction Implementation

[ghstack-poisoned]

edmundw314 committed:
Update on "Bias Correction Implementation"


Differential Revision: [D22661503](https://our.internmc.facebook.com/intern/diff/D22661503)

[ghstack-poisoned]

edmundw314 committed:
Update on "Bias Correction Implementation"


Differential Revision: [D22661503](https://our.internmc.facebook.com/intern/diff/D22661503)

[ghstack-poisoned]

edmundw314 committed:
Update on "Bias Correction Implementation"


Differential Revision: [D22661503](https://our.internmc.facebook.com/intern/diff/D22661503)

[ghstack-poisoned]

edmundw314 committed:
Update on "Bias Correction Implementation"


Differential Revision: [D22661503](https://our.internmc.facebook.com/intern/diff/D22661503)

[ghstack-poisoned]

edmundw314 committed:
Update on "Bias Correction Implementation"


Differential Revision: [D22661503](https://our.internmc.facebook.com/intern/diff/D22661503)

[ghstack-poisoned]

edmundw314 committed:
Update on "Bias Correction Implementation"


Differential Revision: [D22661503](https://our.internmc.facebook.com/intern/diff/D22661503)

[ghstack-poisoned]

edmundw314 committed:
Update on "Bias Correction Implementation"


Differential Revision: [D22661503](https://our.internmc.facebook.com/intern/diff/D22661503)

[ghstack-poisoned]

edmundw314 committed:
Update on "Bias Correction Implementation"


Differential Revision: [D22661503](https://our.internmc.facebook.com/intern/diff/D22661503)

[ghstack-poisoned]

edmundw314 committed:
Update on "Bias Correction Implementation"


Differential Revision: [D22661503](https://our.internmc.facebook.com/intern/diff/D22661503)

[ghstack-poisoned]

edmundw314 committed:
Update on "Bias Correction Implementation"


Differential Revision: [D22661503](https://our.internmc.facebook.com/intern/diff/D22661503)

[ghstack-poisoned]

edmundw314 committed:
Update on "Bias Correction Implementation"


Differential Revision: [D22661503](https://our.internmc.facebook.com/intern/diff/D22661503)

[ghstack-poisoned]

z-a-f commented on torch/quantization/_correct_bias.py:
```diff
@@ -0,0 +1,292 @@
+    remove_hooks(quantized_model)
+    remove_shadow(quantized_model)
+
+def parallel_bias_correction(float_model, quantized_model, img_data, white_list=_supported_modules):
```
I think the naming is slightly incorrect. This is not really parallel -- it is still sequential, it's just batched

z-a-f commented on torch/quantization/_correct_bias.py:
```diff
@@ -0,0 +1,292 @@
+                error_matrix = float_weight - quantized_weight
+
+                # bias correction logic
+                if type(quantized_submodule) in [nn.Conv2d, nnq.Conv2d]:
+                    sum_kernels = torch.sum(error_matrix, (2, 3))  # c_o, c_i
+                    expected_c_i = torch.mean(expected_input[key]['float'], (0, 2, 3))  # c_i
+                    expected_c_i = expected_c_i.reshape(expected_c_i.size()[0], 1)  # flipping into column vector
+                    expected_c_o = torch.matmul(sum_kernels, expected_c_i)  # c_o
+                    expected_c_o = expected_c_o.squeeze(1)
+                elif type(quantized_submodule) in [nn.Linear, nnq.Linear]:
+                    expected_c_o = torch.mean(error_matrix, 1)
```
Is there a paper reference?

z-a-f reviewed (changes requested):
There is one major concern (deleting entries within a loop). Other ones are just nits. Also, linting :)

jerryzh168 commented on test/quantization/test_bias_correction.py:
```diff
@@ -0,0 +1,91 @@
+)
+
+class TestBiasCorrection(QuantizationTestCase):
+    def computeSqnr(self, x, y):
```
guess we are using this in `TestCase`: https://github.com/pytorch/pytorch/blob/master/torch/testing/_internal/common_quantization.py#L160, and here it is following that. maybe we should make theses more consistent

jerryzh168 commented on torch/quantization/_correct_bias.py:
```diff
@@ -0,0 +1,292 @@
+    param = getattr(module, attr, None)
+    if isinstance(param, nn.Parameter):
+        return param
+    elif param is not None:
```
you can check if the param is callable here I think

jerryzh168 commented on torch/quantization/_correct_bias.py:
```diff
@@ -0,0 +1,292 @@
+            self.sum = x
+            self.count = 1
+        else:
+            torch.Tensor.add_(self.sum, x)
```
this can just be `self.sum += x`

jerryzh168 commented on torch/quantization/_correct_bias.py:
```diff
@@ -0,0 +1,292 @@
+        correct_quantized_bias(output_logger, input_logger, self.float_module, self.quantized_module)
+        return a
+
+def add_shadow(float_model, quantized_model, white_list=_supported_modules):
```
is this copying the code from numeric suite?

jerryzh168 commented on torch/quantization/_correct_bias.py:
```diff
@@ -0,0 +1,292 @@
+        correct_quantized_bias(output_logger, input_logger, self.float_module, self.quantized_module)
+        return a
+
+def add_shadow(float_model, quantized_model, white_list=_supported_modules):
```
why can't we use numeric suite to add shadow module?

raghuramank100 replied:
Yes, you should be calling: https://github.com/pytorch/pytorch/blob/master/torch/quantization/_numeric_suite.py#L307

raghuramank100 commented on test/quantization/test_bias_correction.py:
```diff
@@ -0,0 +1,91 @@
+        Pn = torch.norm(x - y)
+        return 20 * torch.log10(Ps / Pn)
+
+    def spnrOfBiasCorrecting(self, float_model, bias_correction, img_data):
```
NIT: Should this be Sqnr?

raghuramank100 commented on torch/quantization/_correct_bias.py:
```diff
@@ -0,0 +1,292 @@
+    for key, value in reassign.items():
+        quantized_model._modules[key] = value
+
+def remove_hooks(model, white_list=_supported_modules):
```
As discussed, you do not need this method here, please remove it from this PR.

raghuramank100 commented on torch/quantization/_correct_bias.py:
```diff
@@ -0,0 +1,292 @@
+            self.stats["tensor_val"] = self.sum / self.count
+        return y
+
+class ShadowModule(nn.Module):
```
You need to define a custom logger instead of defining your own shadow module. The shadow module already contains both the float and quantized modules, and calls the logger in its forward: 
https://github.com/pytorch/pytorch/blob/master/torch/quantization/_numeric_suite.py#L258
As a starting point, you can use the default logger that just captures the output tensors for both float and quantized modules. Later you can customize it to directly compute the difference.

raghuramank100 commented on torch/quantization/_correct_bias.py:
```diff
@@ -0,0 +1,292 @@
+    quantized_model.qconfig = qconfig_debug
+    # calling prepare with prehook param, and throwing the whitelist to make sure supported modules
+    # get the qconfig and thus post hook
+    torch.quantization.prepare(float_model, inplace=True, white_list=white_list, prehook=MeanLogger)
```
You should just create float and quantized models as usual and then use the numeric suite API to insert shadow modules. You do not need prehook or MeanLogger here.

raghuramank100 commented on torch/quantization/_correct_bias.py:
```diff
@@ -0,0 +1,292 @@
+
+    # bias correction is happening under the hood here
+    with torch.no_grad():
+        quantized_model(stack)
```
For sequential correction, the flow is more complicated:
 - quantize the first module, do bias correction for its output
 - update the bias based on the correction
 - now, fix layer 1, it is quantized for both the float and quantize models
 - quantize layer 2 and do bias correction

jerryzh168 replied:
yeah, this function should be called a different name, it's still "parallel" but with special treatment for conv

raghuramank100 commented on torch/quantization/_correct_bias.py:
```diff
@@ -0,0 +1,292 @@
+                if quantized_weight.is_quantized:
+                    quantized_weight = quantized_weight.dequantize()
+
+                error_matrix = float_weight - quantized_weight
```
The logic here is incorrect. Please start with the simpler flow: use the logger to compute the mean of the difference in activations between the float and quantized module, update the bias with this difference. Lets do experiments to see if the same logic works for conv and linear

raghuramank100 commented on torch/quantization/quantize.py:
```diff
@@ -107,8 +114,13 @@ def add_observer_(module, non_leaf_module_list=None, device=None):
+                    child.register_forward_pre_hook(_observer_forward_pre_hook)
         else:
-            add_observer_(child, non_leaf_module_list, device)
+            add_observer_(child, non_leaf_module_list, device, prehook)
```
This should not be needed. observer itself is a hook, so not sure why we need another hook.

edmundw314 replied:
this is because the current implementation here is using the expected input for bias correction (stuff from appendix B), the second hook is to grab the inputs where as the first hook is to get the output

trying to refactor the bias correction logic to not need the appendix B stuff (getting rid of that will get rid of the need for the prehook)

raghuramank100 reviewed (changes requested):
Please see comments

edmundw314 committed:
Update on "Bias Correction Implementation"


Differential Revision: [D22661503](https://our.internmc.facebook.com/intern/diff/D22661503)

[ghstack-poisoned]

jerryzh168 commented on torch/quantization/_correct_bias.py:
```diff
@@ -0,0 +1,292 @@
+        else:
+            stack = torch.cat((stack, data[0]))
+
+    # bias correction is happening under the hood here
```
bias correction should be an explicit step, it should not happen under the hood

jerryzh168 replied:
i see why you are calling this sequential now. this is not what we expect for sequential, please see Raghu's comment for the steps for sequential.

jerryzh168 commented on torch/quantization/_correct_bias.py:
```diff
@@ -0,0 +1,292 @@
+                error_matrix = float_weight - quantized_weight
+
+                # bias correction logic
+                if type(quantized_submodule) in [nn.Conv2d, nnq.Conv2d]:
```
why are you doing bias correction for `nn.Conv2d` and `nn.Linear`?

edmundw314 committed:
Update on "Bias Correction Implementation"


Differential Revision: [D22661503](https://our.internmc.facebook.com/intern/diff/D22661503)

[ghstack-poisoned]

edmundw314 committed:
Update on "Bias Correction Implementation"


Differential Revision: [D22661503](https://our.internmc.facebook.com/intern/diff/D22661503)

[ghstack-poisoned]

edmundw314 committed:
Update on "Bias Correction Implementation"


Differential Revision: [D22661503](https://our.internmc.facebook.com/intern/diff/D22661503)

[ghstack-poisoned]

edmundw314 committed:
Update on "Bias Correction Implementation"


Differential Revision: [D22661503](https://our.internmc.facebook.com/intern/diff/D22661503)

[ghstack-poisoned]

jerryzh168 commented on torch/quantization/_correct_bias.py:
```diff
@@ -0,0 +1,121 @@
+    gives a function that will give you the raw tensor, this function takes care of that logic
+    '''
+    param = getattr(module, attr, None)
+    if isinstance(param, nn.Parameter) or isinstance(param, torch.Tensor):
+        return param
+    elif callable(param):
+        return param()
```
I mean
```python
if callable(param):
  return param()
else:
  return param
```

jerryzh168 commented on test/quantization/test_bias_correction.py:
```diff
@@ -0,0 +1,98 @@
+                if artificial_submodule in _supported_modules:
+                    if artificial_bias.is_quantized:
+                        artificial_bias = artificial_bias.dequantize()
+                    self.assertTrue(self.computeSqnr(float_bias, artificial_bias) > 35,
```
compute_sqnr?

jerryzh168 commented on test/quantization/test_bias_correction.py:
```diff
@@ -0,0 +1,98 @@
+import torch.nn as nn
+from torch.testing._internal.common_quantization import QuantizationTestCase
+from torch.quantization import QuantStub, DeQuantStub
+import torch.quantization._correct_bias as _correct_bias
```
nit: I think you can just have `_` for the filename and remove `_` for the function name so that we don't need to change the callsite when we make this public

jerryzh168 commented on test/quantization/test_bias_correction.py:
```diff
@@ -0,0 +1,98 @@
+                artificial_bias = _correct_bias.get_param(artificial_submodule, 'bias')
+
+                if artificial_submodule in _supported_modules:
+                    if artificial_bias.is_quantized:
```
we don't quantize bias I think, why is this quantized?

jerryzh168 commented on torch/quantization/_correct_bias.py:
```diff
@@ -0,0 +1,121 @@
+_supported_modules = {nn.Linear, nn.Conv2d}
+_supported_modules_quantized = {nnq.Linear, nnq.Conv2d}
+
+def get_module(model, name):
```
I think you can call `model.named_modules()` to get a dictionary of name -> submodules and just use that dictionary to get the submodule.

jerryzh168 commented on torch/quantization/_correct_bias.py:
```diff
@@ -0,0 +1,121 @@
+        current = current._modules[subname]
+    return current
+
+def get_parent_module(model, name):
```
this can be come get_parent_name.
I have some code from GraphModule that will be useful here
```python
# turn foo.bar -> ['foo', 'bar']
def _parent_name(target):
    r = target.rsplit('.', 1)
    if len(r) == 1:
        return '', r[0]
    else:
        return r[0], r[1]
```

jerryzh168 commented on torch/quantization/_correct_bias.py:
```diff
@@ -0,0 +1,121 @@
+        self.float_sum = None
+        self.quant_sum = None
+
+    def forward(self, x, y):
```
looks like x must be quantized and y must be float, please specify this in the comment/docstring

jerryzh168 commented on torch/quantization/_correct_bias.py:
```diff
@@ -0,0 +1,121 @@
+        self.quant_sum = None
+
+    def forward(self, x, y):
+        if len(x) > 1:
```
also what is this?

jerryzh168 commented on torch/quantization/_correct_bias.py:
```diff
@@ -0,0 +1,121 @@
+        if type(submodule) in _supported_modules_quantized:
+            biased_modules[name] = submodule
+
+    for biased_module in biased_modules:
```
this is the name of the biased module, please name it accordingly

jerryzh168 commented on torch/quantization/_correct_bias.py:
```diff
@@ -0,0 +1,121 @@
+        quantized_submodule = get_module(quantized_model, biased_module)
+        bias = get_param(quantized_submodule, 'bias')
+        if bias is not None:
+            ns.prepare_model_with_stubs(float_model, quantized_model, _supported_modules, MeanShadowLogger)
```
is float_model corresponding to `x` and quantizd_model corresponding to `y` in the forward function of `MeanShadowLogger`?

jerryzh168 commented on torch/quantization/_correct_bias.py:
```diff
@@ -0,0 +1,121 @@
+            for data in img_data:
+                quantized_model(data[0])
+                count += 1
+                if count % 10 == 0:
+                    print('.', end='')  # keeps devserver open
```
is this necessary?

jerryzh168 commented on torch/quantization/_correct_bias.py:
```diff
@@ -0,0 +1,121 @@
+            expected_error = torch.mean(quantization_error, dims)
+
+            updated_bias = bias.data - expected_error
+            updated_bias = updated_bias.reshape(bias.data.size())
```
do we need this rehsape?

jerryzh168 reviewed (changes requested):
Looks good overall, please see comments inline.

edmundw314 committed:
Update on "Bias Correction Implementation"


Differential Revision: [D22661503](https://our.internmc.facebook.com/intern/diff/D22661503)

[ghstack-poisoned]

edmundw314 committed:
Update on "Bias Correction Implementation"


Differential Revision: [D22661503](https://our.internmc.facebook.com/intern/diff/D22661503)

[ghstack-poisoned]

edmundw314 committed:
Update on "Bias Correction Implementation"


Differential Revision: [D22661503](https://our.internmc.facebook.com/intern/diff/D22661503)

[ghstack-poisoned]

edmundw314 committed:
Update on "Bias Correction Implementation"


Differential Revision: [D22661503](https://our.internmc.facebook.com/intern/diff/D22661503)

[ghstack-poisoned]

jerryzh168 commented on torch/quantization/_correct_bias.py:
```diff
@@ -0,0 +1,121 @@
+def get_module(model, name):
+    ''' Given name of submodule, this function grabs the submodule from given model
+    '''
+    return {name: submodule for name, submodule in model.named_modules()}[name]
```
you can just do model.named_modules()[name]

edmundw314 replied:
model.named_modules() returns a generator not a dictionary, you can't index into it with a key
i can flip this to 
for module_name, submodule in model.named_modules():
    if name == module_name:
         return submodule
return None

jerryzh168 replied:
hmm, maybe call `dict(model.named_modules())`?

jerryzh168 commented on torch/quantization/_correct_bias.py:
```diff
@@ -0,0 +1,121 @@
+        return param()
+    else:
+        return param
+    return None
```
this line can be removed now

jerryzh168 replied:
this is still valid

jerryzh168 commented on torch/quantization/_correct_bias.py:
```diff
@@ -0,0 +1,121 @@
+            # math for expected_error
+            quantization_error = quant_data - float_data
+            dims = list(range(quantization_error.dim()))
+            dims.remove(1)
```
please add comment for this, we are removing the `output_channel` dimension

jerryzh168 commented on torch/quantization/_correct_bias.py:
```diff
@@ -0,0 +1,121 @@
+            bias.data = updated_bias
+
+            # Removing shadows from model, needed to prevent nesting of shadow modules
+            for name, submodule in quantized_model.named_modules():
```
is this needed?

jerryzh168 replied:
why do we have nested shadow modules? can this be addressed in the way you use numeric suite?

edmundw314 committed:
Update on "Bias Correction Implementation"


Differential Revision: [D22661503](https://our.internmc.facebook.com/intern/diff/D22661503)

[ghstack-poisoned]

edmundw314 committed:
Update on "Bias Correction Implementation"


Differential Revision: [D22661503](https://our.internmc.facebook.com/intern/diff/D22661503)

[ghstack-poisoned]

edmundw314 committed:
Update on "Bias Correction Implementation"


Differential Revision: [D22661503](https://our.internmc.facebook.com/intern/diff/D22661503)

[ghstack-poisoned]

jerryzh168 commented on torch/quantization/_correct_bias.py:
```diff
@@ -0,0 +1,130 @@
+        self.float_sum = None
+        self.quant_sum = None
+
+def bias_correction(float_model, quantized_model, img_data, white_list=_supported_modules_quantized, neval_batches=None):
```
please change `white_list` to `supported_list`

jerryzh168 reviewed (approved):
Had a few comments inline. Also, please click `Resolve Conversation` after the comment is resolved

edmundw314 committed:
Update on "Bias Correction Implementation"


Differential Revision: [D22661503](https://our.internmc.facebook.com/intern/diff/D22661503)

[ghstack-poisoned]

edmundw314 committed:
Update on "Bias Correction Implementation"


Differential Revision: [D22661503](https://our.internmc.facebook.com/intern/diff/D22661503)

[ghstack-poisoned]

edmundw314 committed:
Update on "Bias Correction Implementation"


Differential Revision: [D22661503](https://our.internmc.facebook.com/intern/diff/D22661503)

[ghstack-poisoned]

edmundw314 committed:
Update on "Bias Correction Implementation"


Differential Revision: [D22661503](https://our.internmc.facebook.com/intern/diff/D22661503)

[ghstack-poisoned]

edmundw314 committed:
Update on "Bias Correction Implementation"


Differential Revision: [D22661503](https://our.internmc.facebook.com/intern/diff/D22661503)

[ghstack-poisoned]
